t/basic.t

```diff
@@ -387,6 +387,57 @@ subtest "multicalls" => sub {
   );
 };
 
+subtest 'request validation' => sub {
+  my $uri = $jmap_tester->api_uri;
+
+  subtest "must be post" => sub {
+    my $res = $jmap_tester->ua->get($uri, Content => "{}");
+
+    is($res->code, 405, 'got a 405 response');
+    jcmp_deeply(
+      decode_json($res->decoded_content),
+      { error => "Method not allowed" },
+      'got correct error',
+    );
+  };
+
+  subtest "must be application/json" => sub {
+    my $res = $jmap_tester->ua->post(
+      $uri,
+      'Content-Type' => "junk",
+      Content        => "{}"
+    );
+
+    is($res->code, 415, 'got a 415 response');
+    jcmp_deeply(
+      decode_json($res->decoded_content),
+      { error => "Invalid content-type, must be application/json or application/json;charset=utf-8" },
+      'got correct error',
+    );
+  };
+
+  subtest "other params on application/json okay" => sub {
```

Comment:
This isn't actually what the error message says (which is that it must be application/json or /application/json;charset=utf-8). I'd probably  just make the error message "must be application/json".

None of this actually matters, though because it's an error message no reasonable person will ever see.

Reply:
I didn't want to be too strict as I wasn't sure if other clients were passing weird things in that will work okay, but I wanted the error message to suggest what we reasonably want to accept